301-metadata-repositories.md

```diff
@@ -0,0 +1,307 @@
+
+#### Administrators
+
+The four-top level TUF `root`, `timestamp`, `snapshot`, and `targets` roles SHOULD be controlled by the repository _administrators_, who develop and maintain the repository's hardware and software.
+
+The `root` role distributes, revokes, and replaces the public keys for all four top-level roles. It SHOULD use _offline keys_, or signing keys kept off the Internet (e.g., on trusted hardware in cold storage). It SHOULD also use a threshold (m, n) of keys, where n is the number of administrators, and m is the number of quorum members that must agree on new `root` metadata. Its metadata SHOULD expire yearly, considering that an offline key ceremony is expensive in terms of time and resources.
+
+The `timestamp` role indicates concisely whether there is any new metadata, bundle, or image on the repository.
+
+The `snapshot` role prevents attackers from tampering with interdependencies between bundles and / or images.
+
+Since metadata for both the `timestamp` and `snapshot` roles could be updated whenever a new bundle/ image is produced at any time, they MAY each use a threshold (1, 1) of _online keys_, or signing keys accessible on-demand by automation for the repository. In fact, they MAY even share the same online key. The metadata for each role MAY expire daily, since it is expected to be updated relatively frequently (e.g., whenever a developer pushes a new bundle / image, or a scheduled job runs to renew the metadata).
+
+In order to achieve [gradual security](300-cnab-security.md#gradual-security) over time, it is RECOMMENDED to use the delegation model described in the rest of this subsection. Following the SHOULD recommendations, if either images or bundles are first signed using only TUF, then level 1a or 1b security is provided respectively. If both are signed using TUF, then level 2 security is provided. Building on top of that, following the MAY recommendations, if either images or bundles are additionally signed using in-toto, then level 3a or 3b security is provided. Finally, if both are signed using TUF and in-toto, then level 4 security is provided.
+
+Instead of signing bundles and images itself, the `targets` role SHOULD _delegate_ bundles and images to the `bundles` and `images` roles respectively. With a delegation, a _delegator_ binds a threshold of public keys, which belongs to a _delegatee_, to a _path pattern_, which specifies bundles or images (and in-toto metadata, if any) that the delegatee may be trusted with if the delegator has not signed for them.
+
+More concretely, the `targets` role SHOULD make two backtracking delegations. First, it should delegate all bundles (e.g., `['example.com/example-org/*:*', 'bundles/*']`) to the `bundles` role. Second, it SHOULD delegate all images (e.g., `['example.com/example-org/*:*', 'images/*']`) to the `images` role. In the interest of space, we will discuss only bundles going forward, but note that exactly the same principles apply to images.
```

Comment:
I like this section, and I think it is really important to have it in this document - but would it make more sense to first show what it would mean to have the `targets` role directly sign bundles and see the disadvantages?

Then, organizations that are comfortable with the drawbacks could choose to use it directly, and those who need the additional delegations would better understand why they need them.

Reply:
We could, but my concern is: how many key mgmt models should we discuss? The more we discuss, the longer the document, the less people are likely to read, the more confusing they will find it, and the more we will have to further subdivide pieces. Consider how this would add to the present proposed levels of gradual security:

Level 0: no TUF or in-toto.
Level 1a: TUF for bundles.
Level 1b: TUF for images.
Level 2: TUF for bundles and images.
Level 2b: TUF + in-toto for bundles. TUF for images.
Level 2c: TUF for bundles. TUF + in-toto for images.
Level 3: TUF + in-toto for bundles and images.

Presently, these levels do not account for whether critical signing keys are online or offline (which affects security), and that is highly specific for the key management model.

Don't get me wrong, I think there is a sweet spot somewhere, something like PEP [458](https://www.python.org/dev/peps/pep-0458/) vs [480](https://www.python.org/dev/peps/pep-0480/). We should agree on them before going further. What do you think?